File: src/backend/tests/c-lisp/kernel/runtime/MMult_1x4_4.c
void add_dot1x4(int k, float* a, int lda, float* b, int ldb, float* c, int ldc) {
    c[0]       = 0.;
    c[1 * ldc] = 0.;
    c[2 * ldc] = 0.;
    c[3 * ldc] = 0.;

    int p;

    for(p = 0; p < k; p++)
        c[0] += a[p * lda] * b[p];

    for(p = 0; p < k; p++)
        c[1 * ldc] += a[p * lda] * b[p + ldb * 1];

    for(p = 0; p < k; p++)
        c[2 * ldc] += a[p * lda] * b[p + ldb * 2];

    for(p = 0; p < k; p++)
        c[3 * ldc] += a[p * lda] * b[p + ldb * 3];

}       

void kernel(float* a, float* b, float*c, int m, int n, int k) {
    
    int lda = m;
    int ldb = n;
    int ldc = k;
    
    for(int j = 0; j < n; j += 4) 
        for(int i = 0; i < m; i++) 
            add_dot1x4(k, &a[i], lda, &b[j * ldb], ldb, &c[j * ldc + i], ldc);
}

File: src/backend/tests/c-lisp/kernel/kernels_c-lisp/MMult_1x4_4.py
def init_c(c, ldc):
    return [
        ("store", ("ptradd", c, ("add", 0, ("mul", ele, ldc))), 0.0) for ele in range(3)
    ]


def elemental_muladd_1x4(k, a, b, c, lda, ldb, ldc):
    elemental_muladd = [
        (
            "for",
            (("set", "p", 0), ("lt", "p", k), ("set", "p", ("add", "p", 1))),
            (
                "store",
                ("ptradd", c, ("add", 0, ("mul", ele, ldc))),
                (
                    "fadd",
                    ("load", ("ptradd", c, ("add", 0, ("mul", ele, ldc)))),
                    (
                        "fmul",
                        ("load", ("ptradd", a, ("add", 0, ("mul", "p", lda)))),
                        ("load", ("ptradd", b, ("add", "p", ("mul", ele, ldb)))),
                    ),
                ),
            ),
        )
        for ele in range(4)
    ]

    return elemental_muladd
